pkg/kopia/cli/internal/opts/cache_opts.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2024 The Kanister Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package opts
+
+import (
+	"strconv"
+
+	"github.com/kanisterio/kanister/pkg/kopia/cli/args"
+	"github.com/kanisterio/safecli/command"
+)
+
+const (
+	defaultCacheDirectory = "/tmp/kopia-cache"
+)
+
+// CacheDirectory creates a new cache directory option with a given directory.
+// If the directory is empty, the default cache directory is used.
+func CacheDirectory(dir string) command.Applier {
+	if dir == "" {
+		dir = defaultCacheDirectory
+	}
+	return command.NewOptionWithArgument("--cache-directory", dir)
```

Comment:
`NewOptionWithArgument` would return noop if the value of the argument is `""`.
I wonder if we have any places in the CLI where an empty argument makes logical sense to supply?
Maybe not in this particular case, but there could be some argument somewhere.

Reply:
I haven't seen such a case. An option without an argument is just an option, and we should use `NewOption`.

Reply:
I was talking about deliberately passing an empty string. Since in shell missing argument and argument with an empty string as a value is not the same thing, but `NewOptionWithArgument` has an implicit assumption that it is.
Maybe we could make it explicit somehow?

Reply:
If we remove the specified code from [https://github.com/kanisterio/safecli/command/option_arg.go](https://github.com/kanisterio/safecli/blob/38b6544f3590512f0da53778d152d7bbee8b7c4e/command/option_arg.go#L44-L46) 

```go
...
	} else if arg == "" {
		return noopArgument{}
	}
...
```
then we will generate an option as `--option=` without any value on the right side and this is usually considered a bad argument. We can simulate this approach by just using `NewOption("--option=")` for the empty argument.

I just checked `argparse` (python argument parser) and this function allows you to do it quite well. The decision about whether this is an error should be made on the user side.

```python
# args.py
import argparse

parser = argparse.ArgumentParser(description='cli tool')
parser.add_argument('--option', type=str, default="", nargs='?', help='optional arg with default empty string')

args = parser.parse_args()
print(f"Option value: '{args.option}'")
```

```bash
$ python args.py --option=
Option value: ''
```

You're probably right, we shouldn't prohibit the user from stepping on the rake if he wants to. :)

Let's remove [noopArgument code](https://github.com/kanisterio/safecli/blob/38b6544f3590512f0da53778d152d7bbee8b7c4e/command/option_arg.go#L44-L46)? WDYT?

Reply:
The problem I see there is that it will add `--option=` for non-initialized strings, because for golang an empty string and non-initialized string is the same thing. This is essentially an inconsistency between golang and shell arguments. So if we keep it implicit we still have a problem.
We can make it explicit by maybe providing an option to `NewOptionWithArgument` or making something like `NewOptionWithNonEmptyArgument` to make it clear if you use it you can't pass `""` as an argument.

Reply:
If we remove the [noopArgument code](https://github.com/kanisterio/safecli/blob/38b6544f3590512f0da53778d152d7bbee8b7c4e/command/option_arg.go#L44-L46), we make it **explicit**. The user will receive exactly what they pass in, and we leave the decision about how to handle the empty argument up to the user, similar to the `CacheDirectory` option above.

Adding `isArgRequired` to `NewOptionWithArgument` or creating a new `NewOptionWithNonEmptyArgument` will  pollute `safecli` (we would have to add a similar method for redacted NewXXX functions) IMHO. 

It is also unclear what to do if `isArgRequired` is `true` and the `arg` is empty. Should we ignore the argument or return an error? There's too much logic involved. :thinking:

Reply:
Ok. Fair enough. This means if we want to treat non-initialized strings as no argument we need to move noop logic a level higher.

Reply:
Right, and we already have `NewNoopArgument`, which can help build this logic.

```go
// OptWithOptionalArgAndDefault("") => --opt=default
// OptWithOptionalArgAndDefault("custom") => --opt=custom
func OptWithOptionalArgAndDefault(arg string) command.Applier {
	if arg == "" {
		arg = "default"
	}
	return command.NewOptionWithArgument("--opt", dir)
}

// OptWithOptionalArgAndNoop("") => nothing will be generated
// OptWithOptionalArgAndNoop("custom") => --opt=custom
func OptWithOptionalArgAndNoop(opt string) command.Applier {
	if opt == "" {
		return command.NewNoopArgument()
	}
	return command.NewOptionWithArgument("--opt", opt)
}

// OptWithRequiredArgAndError("") => error
// OptWithRequiredArgAndError("custom") => --opt=custom
func OptWithRequiredArgAndError(opt string) command.Applier {
	if opt == "" {
		return command.NewErrorArgument(fmt.Errorf("opt is required"))
	}
	return command.NewOptionWithArgument("--opt", opt)
}
```